File: fiks-meldingsformidler/src/main/java/no/difi/meldingsutveksling/ks/fiksio/FiksIoConfig.java
package no.difi.meldingsutveksling.ks.fiksio;

import lombok.RequiredArgsConstructor;
import no.difi.meldingsutveksling.config.IntegrasjonspunktProperties;
import no.difi.move.common.cert.KeystoreHelper;
import no.ks.fiks.io.client.FiksIOKlient;
import no.ks.fiks.io.client.FiksIOKlientFactory;
import no.ks.fiks.io.client.konfigurasjon.*;
import no.ks.fiks.io.client.model.KontoId;
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

import java.security.KeyStore;
import java.util.UUID;

@Configuration
@RequiredArgsConstructor
@ConditionalOnProperty(name = {"difi.move.feature.enableDPFIO"}, havingValue = "true")
public class FiksIoConfig {

    private final IntegrasjonspunktProperties props;
    private final KeystoreHelper keystoreHelper;

    @Bean
    public FiksIOKlient fiksIoKlient() {
        try {
            KeyStore oidcKeystore = KeyStore.getInstance(props.getOidc().getKeystore().getType());
            oidcKeystore.load(props.getOidc().getKeystore().getPath().getInputStream(), props.getOidc().getKeystore().getPassword().toCharArray());
            UUID kontoId = UUID.fromString(props.getFiks().getIo().getKontoId());

            FiksIOKonfigurasjon fiksIOConfig = FiksIOKonfigurasjon.builder()
                .amqpKonfigurasjon(AmqpKonfigurasjon.builder()
                    .host(props.getFiks().getIo().getHost())
                    .build())
                .fiksApiKonfigurasjon(FiksApiKonfigurasjon.builder()
                    .host(props.getFiks().getIo().getApiHost())
                    .port(443)
                    .scheme("https")
                    .build())
                .kontoKonfigurasjon(KontoKonfigurasjon.builder()
                    .kontoId(new KontoId(kontoId))
                    .privatNokkel(keystoreHelper.loadPrivateKey())
                    .build())
                .fiksIntegrasjonKonfigurasjon(FiksIntegrasjonKonfigurasjon.builder()
                    .integrasjonId(UUID.fromString(props.getFiks().getIo().getIntegrasjonsId()))
                    .integrasjonPassord(props.getFiks().getIo().getIntegrasjonsPassord())
                    .idPortenKonfigurasjon(IdPortenKonfigurasjon.builder()
                        .accessTokenUri(props.getOidc().getUrl().toString())
                        .idPortenAudience(props.getOidc().getAudience())
                        .klientId(props.getOidc().getClientId())
                        .build())
                    .build())
                .virksomhetssertifikatKonfigurasjon(VirksomhetssertifikatKonfigurasjon.builder()
                    .keyStore(oidcKeystore)
                    .keyAlias(props.getOidc().getKeystore().getAlias())
                    .keyPassword(props.getOidc().getKeystore().getPassword())
                    .keyStorePassword(props.getOidc().getKeystore().getPassword())
                    .build())
                .build();

            return new FiksIOKlientFactory(fiksIOConfig).build();
        } catch (Exception e) {
            throw new IllegalStateException("Failed to initialize FiksIOKlient", e);
        }
    }

}

File: fiks-meldingsformidler/src/main/kotlin/no/difi/meldingsutveksling/ks/fiksio/FiksIoConfig.kt
package no.difi.meldingsutveksling.ks.fiksio

import no.difi.meldingsutveksling.IntegrasjonspunktNokkel
import no.difi.meldingsutveksling.config.IntegrasjonspunktProperties
import no.ks.fiks.io.client.FiksIOKlient
import no.ks.fiks.io.client.FiksIOKlientFactory
import no.ks.fiks.io.client.konfigurasjon.*
import no.ks.fiks.io.client.model.KontoId
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import java.security.KeyStore
import java.util.*

@Configuration
@ConditionalOnProperty(name = ["difi.move.feature.enableDPFIO"], havingValue = "true")
open class FiksIoConfig(private val props: IntegrasjonspunktProperties,
                        private val ipNokkel: IntegrasjonspunktNokkel) {

    @Bean
    open fun fiksIoKlient(): FiksIOKlient {
        val oidcKeystore = KeyStore.getInstance(props.oidc.keystore.type)
        oidcKeystore.load(props.oidc.keystore.path.inputStream, props.oidc.keystore.password.toCharArray())
        val kontoId: UUID = UUID.fromString(props.fiks.io.kontoId)

        val fiksIOConfig = FiksIOKonfigurasjon.builder()
                .amqpKonfigurasjon(AmqpKonfigurasjon.builder()
                        .host(props.fiks.io.host)
                        .build())
                .fiksApiKonfigurasjon(FiksApiKonfigurasjon.builder()
                        .host(props.fiks.io.apiHost)
                        .port(443)
                        .scheme("https")
                        .build())
                .kontoKonfigurasjon(KontoKonfigurasjon.builder()
                        .kontoId(KontoId(kontoId))
                        .privatNokkel(ipNokkel.loadPrivateKey())
                        .build())
                .fiksIntegrasjonKonfigurasjon(FiksIntegrasjonKonfigurasjon.builder()
                        .integrasjonId(UUID.fromString(props.fiks.io.integrasjonsId))
                        .integrasjonPassord(props.fiks.io.integrasjonsPassord)
                        .idPortenKonfigurasjon(IdPortenKonfigurasjon.builder()
                                .accessTokenUri(props.oidc.url.toString())
                                .idPortenAudience(props.oidc.audience)
                                .klientId(props.oidc.clientId)
                                .build())
                        .build())
                .virksomhetssertifikatKonfigurasjon(VirksomhetssertifikatKonfigurasjon.builder()
                        .keyStore(oidcKeystore)
                        .keyAlias(props.oidc.keystore.alias)
                        .keyPassword(props.oidc.keystore.password)
                        .keyStorePassword(props.oidc.keystore.password)
                        .build())
                .build()
        return FiksIOKlientFactory(fiksIOConfig).build()
    }
}
